
[Skip to content](#start-of-content)

## Navigation Menu

Toggle navigation

[Sign in](/login?return_to=https%3A%2F%2Fgithub.com%2Fpython%2Fcpython%2Fissues%2F114572)

* Product

  + [GitHub Copilot
    Write better code with AI](https://github.com/features/copilot)
  + [Security
    Find and fix vulnerabilities](https://github.com/features/security)
  + [Actions
    Automate any workflow](https://github.com/features/actions)
  + [Codespaces
    Instant dev environments](https://github.com/features/codespaces)
  + [Issues
    Plan and track work](https://github.com/features/issues)
  + [Code Review
    Manage code changes](https://github.com/features/code-review)
  + [Discussions
    Collaborate outside of code](https://github.com/features/discussions)
  + [Code Search
    Find more, search less](https://github.com/features/code-search)

  Explore
  + [All features](https://github.com/features)
  + [Documentation](https://docs.github.com)
  + [GitHub Skills](https://skills.github.com)
  + [Blog](https://github.blog)
* Solutions

  By company size
  + [Enterprises](https://github.com/enterprise)
  + [Small and medium teams](https://github.com/team)
  + [Startups](https://github.com/enterprise/startups)
  By use case
  + [DevSecOps](/solutions/use-case/devsecops)
  + [DevOps](/solutions/use-case/devops)
  + [CI/CD](/solutions/use-case/ci-cd)
  + [View all use cases](/solutions/use-case)

  By industry
  + [Healthcare](/solutions/industry/healthcare)
  + [Financial services](/solutions/industry/financial-services)
  + [Manufacturing](/solutions/industry/manufacturing)
  + [Government](/solutions/industry/government)
  + [View all industries](/solutions/industry)

  [View all solutions](/solutions)
* Resources

  Topics
  + [AI](/resources/articles/ai)
  + [DevOps](/resources/articles/devops)
  + [Security](/resources/articles/security)
  + [Software Development](/resources/articles/software-development)
  + [View all](/resources/articles)

  Explore
  + [Learning Pathways](https://resources.github.com/learn/pathways)
  + [White papers, Ebooks, Webinars](https://resources.github.com)
  + [Customer Stories](https://github.com/customer-stories)
  + [Partners](https://partner.github.com)
  + [Executive Insights](https://github.com/solutions/executive-insights)
* Open Source

  + [GitHub Sponsors
    Fund open source developers](/sponsors)
  + [The ReadME Project
    GitHub community articles](https://github.com/readme)
  Repositories
  + [Topics](https://github.com/topics)
  + [Trending](https://github.com/trending)
  + [Collections](https://github.com/collections)
* Enterprise

  + [Enterprise platform
    AI-powered developer platform](/enterprise)
  Available add-ons
  + [Advanced Security
    Enterprise-grade security features](https://github.com/enterprise/advanced-security)
  + [GitHub Copilot
    Enterprise-grade AI features](/features/copilot#enterprise)
  + [Premium Support
    Enterprise-grade 24/7 support](/premium-support)
* [Pricing](https://github.com/pricing)

Search or jump to...

# Search code, repositories, users, issues, pull requests...

Search

Clear

[Search syntax tips](https://docs.github.com/search-github/github-code-search/understanding-github-code-search-syntax)

# Provide feedback

We read every piece of feedback, and take your input very seriously.

Include my email address so I can be contacted

  Cancel

 Submit feedback

# Saved searches

## Use saved searches to filter your results more quickly

Name

Query

To see all available qualifiers, see our [documentation](https://docs.github.com/search-github/github-code-search/understanding-github-code-search-syntax).

  Cancel

 Create saved search

[Sign in](/login?return_to=https%3A%2F%2Fgithub.com%2Fpython%2Fcpython%2Fissues%2F114572)

[Sign up](/signup?ref_cta=Sign+up&ref_loc=header+logged+out&ref_page=%2F%3Cuser-name%3E%2F%3Crepo-name%3E%2Fvoltron%2Fissues_fragments%2Fissue_layout&source=header-repo&source_repo=python%2Fcpython)
Reseting focus

You signed in with another tab or window. Reload to refresh your session.
You signed out in another tab or window. Reload to refresh your session.
You switched accounts on another tab or window. Reload to refresh your session.

Dismiss alert

{{ message }}

[python](/python)
/
**[cpython](/python/cpython)**
Public

* [Notifications](/login?return_to=%2Fpython%2Fcpython) You must be signed in to change notification settings
* [Fork
  30.8k](/login?return_to=%2Fpython%2Fcpython)
* [Star
   64.7k](/login?return_to=%2Fpython%2Fcpython)

* [Code](/python/cpython)
* [Issues
  5k+](/python/cpython/issues)
* [Pull requests
  1.8k](/python/cpython/pulls)
* [Actions](/python/cpython/actions)
* [Projects
  28](/python/cpython/projects)
* [Security](/python/cpython/security)
* [Insights](/python/cpython/pulse)

Additional navigation options

* [Code](/python/cpython)
* [Issues](/python/cpython/issues)
* [Pull requests](/python/cpython/pulls)
* [Actions](/python/cpython/actions)
* [Projects](/python/cpython/projects)
* [Security](/python/cpython/security)
* [Insights](/python/cpython/pulse)

New issue

**Have a question about this project?** Sign up for a free GitHub account to open an issue and contact its maintainers and the community.

 [Sign up for GitHub](/signup?return_to=%2Fpython%2Fcpython%2Fissues%2Fnew%2Fchoose)

By clicking “Sign up for GitHub”, you agree to our [terms of service](https://docs.github.com/terms) and
[privacy statement](https://docs.github.com/privacy). We’ll occasionally send you account related emails.

Already on GitHub?
[Sign in](/login?return_to=%2Fpython%2Fcpython%2Fissues%2Fnew%2Fchoose)
to your account

[Jump to bottom](#issue-comment-box)

# ssl: Fix locking in cert\_store\_stats and get\_ca\_certs #114572

Closed

[davidben](/davidben) opened this issue
Jan 25, 2024
· 1 comment

Closed

# [ssl: Fix locking in cert\_store\_stats and get\_ca\_certs](#top) #114572

[davidben](/davidben) opened this issue
Jan 25, 2024
· 1 comment

Labels
[type-bug](/python/cpython/labels/type-bug)
An unexpected behavior, bug, or error

## Comments

[![@davidben](https://avatars.githubusercontent.com/u/109690?s=80&v=4)](/davidben)

Copy link

Contributor

### **[davidben](/davidben)** commented [Jan 25, 2024](#issue-2101303751) • edited by bedevere-app bot Loading

| Bug reportBug description: Filing this to attach a PR to.  cert\_store\_stats and get\_ca\_certs query the SSLContext's X509\_STORE with X509\_STORE\_get0\_objects, but reading the result requires a lock. See [openssl/openssl#23224](https://github.com/openssl/openssl/pull/23224) for details. CPython versions tested on: CPython main branch Operating systems tested on: Linux, macOS Linked PRs  * [gh-114572: Fix locking in cert\_store\_stats and get\_ca\_certs #114573](https://github.com/python/cpython/pull/114573) * [[3.12] gh-114572: Fix locking in cert\_store\_stats and get\_ca\_certs (GH-114573) #115547](https://github.com/python/cpython/pull/115547) * [[3.10] gh-114572: Fix locking in cert\_store\_stats and get\_ca\_certs (GH-114573) #115548](https://github.com/python/cpython/pull/115548) * [[3.11] gh-114572: Fix locking in cert\_store\_stats and get\_ca\_certs (GH-114573) #115549](https://github.com/python/cpython/pull/115549) * [[3.9] gh-114572: Fix locking in cert\_store\_stats and get\_ca\_certs #118109](https://github.com/python/cpython/pull/118109) * [[3.8] gh-114572: Fix locking in cert\_store\_stats and get\_ca\_certs #118442](https://github.com/python/cpython/pull/118442) |
| --- |
| The text was updated successfully, but these errors were encountered: |

All reactions

[![@davidben](https://avatars.githubusercontent.com/u/109690?s=40&v=4)](/davidben)
[davidben](/davidben)
added
the
[type-bug](/python/cpython/labels/type-bug)
An unexpected behavior, bug, or error
label
[Jan 25, 2024](#event-11603185932)

[davidben](/davidben)
added a commit
to davidben/cpython
that referenced
this issue
[Jan 25, 2024](#ref-commit-529311e)
[![@davidben](https://avatars.githubusercontent.com/u/109690?s=40&v=4)](/davidben)

`[pythongh-114572](https://github.com/python/cpython/issues/114572)[: Fix locking in cert_store_stats and get_ca_certs](/davidben/cpython/commit/529311e4390a72bf500662ee9567d9113c1d0aec "gh-114572: Fix locking in cert_store_stats and get_ca_certs

cert_store_stats and get_ca_certs query the SSLContext's X509_STORE with
X509_STORE_get0_objects, but reading the result requires a lock. See
https://github.com/openssl/openssl/pull/23224 for details.

Instead, use X509_STORE_get1_objects, newly added in that PR.
X509_STORE_get1_objects does not exist in current OpenSSLs, but we can
polyfill it with X509_STORE_lock and X509_STORE_unlock.")`
…

`[529311e](/davidben/cpython/commit/529311e4390a72bf500662ee9567d9113c1d0aec)`

```
cert_store_stats and get_ca_certs query the SSLContext's X509_STORE with
X509_STORE_get0_objects, but reading the result requires a lock. See
[openssl/openssl#23224](https://github.com/openssl/openssl/pull/23224) for details.

Instead, use X509_STORE_get1_objects, newly added in that PR.
X509_STORE_get1_objects does not exist in current OpenSSLs, but we can
polyfill it with X509_STORE_lock and X509_STORE_unlock.
```

[![@bedevere-app](https://avatars.githubusercontent.com/in/388350?s=40&v=4)](/apps/bedevere-app)
[bedevere-app](/apps/bedevere-app)
bot
mentioned this issue
[Jan 25, 2024](#ref-pullrequest-2101304587)

[gh-114572: Fix locking in cert\_store\_stats and get\_ca\_certs
#114573](/python/cpython/pull/114573)
 Merged

[alex](/alex)
pushed a commit
that referenced
this issue
[Feb 16, 2024](#ref-commit-bce6931)
[![@davidben](https://avatars.githubusercontent.com/u/109690?s=40&v=4)](/davidben)

`[gh-114572](https://github.com/python/cpython/issues/114572)[: Fix locking in cert_store_stats and get_ca_certs (](/python/cpython/commit/bce693111bff906ccf9281c22371331aaff766ab "gh-114572: Fix locking in cert_store_stats and get_ca_certs (#114573)

* gh-114572: Fix locking in cert_store_stats and get_ca_certs

cert_store_stats and get_ca_certs query the SSLContext's X509_STORE with
X509_STORE_get0_objects, but reading the result requires a lock. See
https://github.com/openssl/openssl/pull/23224 for details.

Instead, use X509_STORE_get1_objects, newly added in that PR.
X509_STORE_get1_objects does not exist in current OpenSSLs, but we can
polyfill it with X509_STORE_lock and X509_STORE_unlock.

* Work around const-correctness problem

* Add missing X509_STORE_get1_objects failure check

* Add blurb")[#114573](https://github.com/python/cpython/pull/114573)[)](/python/cpython/commit/bce693111bff906ccf9281c22371331aaff766ab "gh-114572: Fix locking in cert_store_stats and get_ca_certs (#114573)

* gh-114572: Fix locking in cert_store_stats and get_ca_certs

cert_store_stats and get_ca_certs query the SSLContext's X509_STORE with
X509_STORE_get0_objects, but reading the result requires a lock. See
https://github.com/openssl/openssl/pull/23224 for details.

Instead, use X509_STORE_get1_objects, newly added in that PR.
X509_STORE_get1_objects does not exist in current OpenSSLs, but we can
polyfill it with X509_STORE_lock and X509_STORE_unlock.

* Work around const-correctness problem

* Add missing X509_STORE_get1_objects failure check

* Add blurb")`
…

`[bce6931](/python/cpython/commit/bce693111bff906ccf9281c22371331aaff766ab)`

```
* [gh-114572](https://github.com/python/cpython/issues/114572): Fix locking in cert_store_stats and get_ca_certs

cert_store_stats and get_ca_certs query the SSLContext's X509_STORE with
X509_STORE_get0_objects, but reading the result requires a lock. See
[openssl/openssl#23224](https://github.com/openssl/openssl/pull/23224) for details.

Instead, use X509_STORE_get1_objects, newly added in that PR.
X509_STORE_get1_objects does not exist in current OpenSSLs, but we can
polyfill it with X509_STORE_lock and X509_STORE_unlock.

* Work around const-correctness problem

* Add missing X509_STORE_get1_objects failure check

* Add blurb
```

[miss-islington](/miss-islington)
pushed a commit
to miss-islington/cpython
that referenced
this issue
[Feb 16, 2024](#ref-commit-5ecee31)
[![@davidben](https://avatars.githubusercontent.com/u/109690?s=40&v=4)](/davidben) [![@miss-islington](https://avatars.githubusercontent.com/u/31488909?s=40&v=4)](/miss-islington)

`[pythongh-114572](https://github.com/python/cpython/issues/114572)[: Fix locking in cert_store_stats and get_ca_certs (](/miss-islington/cpython/commit/5ecee311d613c8e5dc5510245eda0f9781b5daed "gh-114572: Fix locking in cert_store_stats and get_ca_certs (GH-114573)

* gh-114572: Fix locking in cert_store_stats and get_ca_certs

cert_store_stats and get_ca_certs query the SSLContext's X509_STORE with
X509_STORE_get0_objects, but reading the result requires a lock. See
https://github.com/openssl/openssl/pull/23224 for details.

Instead, use X509_STORE_get1_objects, newly added in that PR.
X509_STORE_get1_objects does not exist in current OpenSSLs, but we can
polyfill it with X509_STORE_lock and X509_STORE_unlock.

* Work around const-correctness problem

* Add missing X509_STORE_get1_objects failure check

* Add blurb
(cherry picked from commit bce693111bff906ccf9281c22371331aaff766ab)

Co-authored-by: David Benjamin <davidben@google.com>")[py…](https://github.com/python/cpython/pull/114573)`
…

`[5ecee31](/miss-islington/cpython/commit/5ecee311d613c8e5dc5510245eda0f9781b5daed)`

```
[…thonGH-114573](https://github.com/python/cpython/pull/114573))

* [pythongh-114572](https://github.com/python/cpython/issues/114572): Fix locking in cert_store_stats and get_ca_certs

cert_store_stats and get_ca_certs query the SSLContext's X509_STORE with
X509_STORE_get0_objects, but reading the result requires a lock. See
[openssl/openssl#23224](https://github.com/openssl/openssl/pull/23224) for details.

Instead, use X509_STORE_get1_objects, newly added in that PR.
X509_STORE_get1_objects does not exist in current OpenSSLs, but we can
polyfill it with X509_STORE_lock and X509_STORE_unlock.

* Work around const-correctness problem

* Add missing X509_STORE_get1_objects failure check

* Add blurb
(cherry picked from commit [bce6931](https://github.com/miss-islington/cpython/commit/bce693111bff906ccf9281c22371331aaff766ab))

Co-authored-by: David Benjamin <davidben@google.com>
```

[![@miss-islington](https://avatars.githubusercontent.com/u/31488909?s=40&u=31ad766412aaeb40a51582d882e34d67f2fae635&v=4)](/miss-islington)
[miss-islington](/miss-islington)
mentioned this issue
[Feb 16, 2024](#ref-pullrequest-2137718864)

[[3.12] gh-114572: Fix locking in cert\_store\_stats and get\_ca\_certs (GH-114573)
#115547](/python/cpython/pull/115547)
 Merged

[miss-islington](/miss-islington)
pushed a commit
to miss-islington/cpython
that referenced
this issue
[Feb 16, 2024](#ref-commit-597d1bf)
[![@davidben](https://avatars.githubusercontent.com/u/109690?s=40&v=4)](/davidben) [![@miss-islington](https://avatars.githubusercontent.com/u/31488909?s=40&v=4)](/miss-islington)

`[pythongh-114572](https://github.com/python/cpython/issues/114572)[: Fix locking in cert_store_stats and get_ca_certs (](/miss-islington/cpython/commit/597d1bfe2acdab8ef1ac30854d33b2f60b901384 "gh-114572: Fix locking in cert_store_stats and get_ca_certs (GH-114573)

* gh-114572: Fix locking in cert_store_stats and get_ca_certs

cert_store_stats and get_ca_certs query the SSLContext's X509_STORE with
X509_STORE_get0_objects, but reading the result requires a lock. See
https://github.com/openssl/openssl/pull/23224 for details.

Instead, use X509_STORE_get1_objects, newly added in that PR.
X509_STORE_get1_objects does not exist in current OpenSSLs, but we can
polyfill it with X509_STORE_lock and X509_STORE_unlock.

* Work around const-correctness problem

* Add missing X509_STORE_get1_objects failure check

* Add blurb
(cherry picked from commit bce693111bff906ccf9281c22371331aaff766ab)

Co-authored-by: David Benjamin <davidben@google.com>")[py…](https://github.com/python/cpython/pull/114573)`
…

`[597d1bf](/miss-islington/cpython/commit/597d1bfe2acdab8ef1ac30854d33b2f60b901384)`

```
[…thonGH-114573](https://github.com/python/cpython/pull/114573))

* [pythongh-114572](https://github.com/python/cpython/issues/114572): Fix locking in cert_store_stats and get_ca_certs

cert_store_stats and get_ca_certs query the SSLContext's X509_STORE with
X509_STORE_get0_objects, but reading the result requires a lock. See
[openssl/openssl#23224](https://github.com/openssl/openssl/pull/23224) for details.

Instead, use X509_STORE_get1_objects, newly added in that PR.
X509_STORE_get1_objects does not exist in current OpenSSLs, but we can
polyfill it with X509_STORE_lock and X509_STORE_unlock.

* Work around const-correctness problem

* Add missing X509_STORE_get1_objects failure check

* Add blurb
(cherry picked from commit [bce6931](https://github.com/miss-islington/cpython/commit/bce693111bff906ccf9281c22371331aaff766ab))

Co-authored-by: David Benjamin <davidben@google.com>
```

This was referenced Feb 16, 2024

[[3.10] gh-114572: Fix locking in cert\_store\_stats and get\_ca\_certs (GH-114573)
#115548](/python/cpython/pull/115548)
 Merged

[[3.11] gh-114572: Fix locking in cert\_store\_stats and get\_ca\_certs (GH-114573)
#115549](/python/cpython/pull/115549)
 Merged

[miss-islington](/miss-islington)
pushed a commit
to miss-islington/cpython
that referenced
this issue
[Feb 16, 2024](#ref-commit-fa5c6d1)
[![@davidben](https://avatars.githubusercontent.com/u/109690?s=40&v=4)](/davidben) [![@miss-islington](https://avatars.githubusercontent.com/u/31488909?s=40&v=4)](/miss-islington)

`[pythongh-114572](https://github.com/python/cpython/issues/114572)[: Fix locking in cert_store_stats and get_ca_certs (](/miss-islington/cpython/commit/fa5c6d1c4b3e6556cddf663d7b36ed7cdbbde18c "gh-114572: Fix locking in cert_store_stats and get_ca_certs (GH-114573)

* gh-114572: Fix locking in cert_store_stats and get_ca_certs

cert_store_stats and get_ca_certs query the SSLContext's X509_STORE with
X509_STORE_get0_objects, but reading the result requires a lock. See
https://github.com/openssl/openssl/pull/23224 for details.

Instead, use X509_STORE_get1_objects, newly added in that PR.
X509_STORE_get1_objects does not exist in current OpenSSLs, but we can
polyfill it with X509_STORE_lock and X509_STORE_unlock.

* Work around const-correctness problem

* Add missing X509_STORE_get1_objects failure check

* Add blurb
(cherry picked from commit bce693111bff906ccf9281c22371331aaff766ab)

Co-authored-by: David Benjamin <davidben@google.com>")[py…](https://github.com/python/cpython/pull/114573)`
…

`[fa5c6d1](/miss-islington/cpython/commit/fa5c6d1c4b3e6556cddf663d7b36ed7cdbbde18c)`

```
[…thonGH-114573](https://github.com/python/cpython/pull/114573))

* [pythongh-114572](https://github.com/python/cpython/issues/114572): Fix locking in cert_store_stats and get_ca_certs

cert_store_stats and get_ca_certs query the SSLContext's X509_STORE with
X509_STORE_get0_objects, but reading the result requires a lock. See
[openssl/openssl#23224](https://github.com/openssl/openssl/pull/23224) for details.

Instead, use X509_STORE_get1_objects, newly added in that PR.
X509_STORE_get1_objects does not exist in current OpenSSLs, but we can
polyfill it with X509_STORE_lock and X509_STORE_unlock.

* Work around const-correctness problem

* Add missing X509_STORE_get1_objects failure check

* Add blurb
(cherry picked from commit [bce6931](https://github.com/miss-islington/cpython/commit/bce693111bff906ccf9281c22371331aaff766ab))

Co-authored-by: David Benjamin <davidben@google.com>
```

[pablogsal](/pablogsal)
pushed a commit
that referenced
this issue
[Feb 20, 2024](#ref-commit-01c37f1)
[![@miss-islington](https://avatars.githubusercontent.com/u/31488909?s=40&u=31ad766412aaeb40a51582d882e34d67f2fae635&v=4)](/miss-islington) [![@davidben](https://avatars.githubusercontent.com/u/109690?s=40&v=4)](/davidben)

`[[3.11]](/python/cpython/commit/01c37f1d0714f5822d34063ca7180b595abf589d "[3.11] gh-114572: Fix locking in cert_store_stats and get_ca_certs (GH-114573) (#115549)

gh-114572: Fix locking in cert_store_stats and get_ca_certs (GH-114573)

* gh-114572: Fix locking in cert_store_stats and get_ca_certs

cert_store_stats and get_ca_certs query the SSLContext's X509_STORE with
X509_STORE_get0_objects, but reading the result requires a lock. See
https://github.com/openssl/openssl/pull/23224 for details.

Instead, use X509_STORE_get1_objects, newly added in that PR.
X509_STORE_get1_objects does not exist in current OpenSSLs, but we can
polyfill it with X509_STORE_lock and X509_STORE_unlock.

* Work around const-correctness problem

* Add missing X509_STORE_get1_objects failure check

* Add blurb
(cherry picked from commit bce693111bff906ccf9281c22371331aaff766ab)

Co-authored-by: David Benjamin <davidben@google.com>") [gh-114572](https://github.com/python/cpython/issues/114572)[: Fix locking in cert_store_stats and get_ca_certs (](/python/cpython/commit/01c37f1d0714f5822d34063ca7180b595abf589d "[3.11] gh-114572: Fix locking in cert_store_stats and get_ca_certs (GH-114573) (#115549)

gh-114572: Fix locking in cert_store_stats and get_ca_certs (GH-114573)

* gh-114572: Fix locking in cert_store_stats and get_ca_certs

cert_store_stats and get_ca_certs query the SSLContext's X509_STORE with
X509_STORE_get0_objects, but reading the result requires a lock. See
https://github.com/openssl/openssl/pull/23224 for details.

Instead, use X509_STORE_get1_objects, newly added in that PR.
X509_STORE_get1_objects does not exist in current OpenSSLs, but we can
polyfill it with X509_STORE_lock and X509_STORE_unlock.

* Work around const-correctness problem

* Add missing X509_STORE_get1_objects failure check

* Add blurb
(cherry picked from commit bce693111bff906ccf9281c22371331aaff766ab)

Co-authored-by: David Benjamin <davidben@google.com>")[G…](https://github.com/python/cpython/pull/114573)`
…

`[01c37f1](/python/cpython/commit/01c37f1d0714f5822d34063ca7180b595abf589d)`

```
[…H-114573](https://github.com/python/cpython/pull/114573)) ([#115549](https://github.com/python/cpython/pull/115549))

[gh-114572](https://github.com/python/cpython/issues/114572): Fix locking in cert_store_stats and get_ca_certs ([GH-114573](https://github.com/python/cpython/pull/114573))

* [gh-114572](https://github.com/python/cpython/issues/114572): Fix locking in cert_store_stats and get_ca_certs

cert_store_stats and get_ca_certs query the SSLContext's X509_STORE with
X509_STORE_get0_objects, but reading the result requires a lock. See
[openssl/openssl#23224](https://github.com/openssl/openssl/pull/23224) for details.

Instead, use X509_STORE_get1_objects, newly added in that PR.
X509_STORE_get1_objects does not exist in current OpenSSLs, but we can
polyfill it with X509_STORE_lock and X509_STORE_unlock.

* Work around const-correctness problem

* Add missing X509_STORE_get1_objects failure check

* Add blurb
(cherry picked from commit [bce6931](https://github.com/python/cpython/commit/bce693111bff906ccf9281c22371331aaff766ab))

Co-authored-by: David Benjamin <davidben@google.com>
```

[pablogsal](/pablogsal)
pushed a commit
that referenced
this issue
[Feb 20, 2024](#ref-commit-37324b4)
[![@miss-islington](https://avatars.githubusercontent.com/u/31488909?s=40&u=31ad766412aaeb40a51582d882e34d67f2fae635&v=4)](/miss-islington) [![@davidben](https://avatars.githubusercontent.com/u/109690?s=40&v=4)](/davidben)

`[[3.10]](/python/cpython/commit/37324b421b72b7bc9934e27aba85d48d4773002e "[3.10] gh-114572: Fix locking in cert_store_stats and get_ca_certs (GH-114573) (#115548)

gh-114572: Fix locking in cert_store_stats and get_ca_certs (GH-114573)

* gh-114572: Fix locking in cert_store_stats and get_ca_certs

cert_store_stats and get_ca_certs query the SSLContext's X509_STORE with
X509_STORE_get0_objects, but reading the result requires a lock. See
https://github.com/openssl/openssl/pull/23224 for details.

Instead, use X509_STORE_get1_objects, newly added in that PR.
X509_STORE_get1_objects does not exist in current OpenSSLs, but we can
polyfill it with X509_STORE_lock and X509_STORE_unlock.

* Work around const-correctness problem

* Add missing X509_STORE_get1_objects failure check

* Add blurb
(cherry picked from commit bce693111bff906ccf9281c22371331aaff766ab)

Co-authored-by: David Benjamin <davidben@google.com>") [gh-114572](https://github.com/python/cpython/issues/114572)[: Fix locking in cert_store_stats and get_ca_certs (](/python/cpython/commit/37324b421b72b7bc9934e27aba85d48d4773002e "[3.10] gh-114572: Fix locking in cert_store_stats and get_ca_certs (GH-114573) (#115548)

gh-114572: Fix locking in cert_store_stats and get_ca_certs (GH-114573)

* gh-114572: Fix locking in cert_store_stats and get_ca_certs

cert_store_stats and get_ca_certs query the SSLContext's X509_STORE with
X509_STORE_get0_objects, but reading the result requires a lock. See
https://github.com/openssl/openssl/pull/23224 for details.

Instead, use X509_STORE_get1_objects, newly added in that PR.
X509_STORE_get1_objects does not exist in current OpenSSLs, but we can
polyfill it with X509_STORE_lock and X509_STORE_unlock.

* Work around const-correctness problem

* Add missing X509_STORE_get1_objects failure check

* Add blurb
(cherry picked from commit bce693111bff906ccf9281c22371331aaff766ab)

Co-authored-by: David Benjamin <davidben@google.com>")[G…](https://github.com/python/cpython/pull/114573)`
…

`[37324b4](/python/cpython/commit/37324b421b72b7bc9934e27aba85d48d4773002e)`

```
[…H-114573](https://github.com/python/cpython/pull/114573)) ([#115548](https://github.com/python/cpython/pull/115548))

[gh-114572](https://github.com/python/cpython/issues/114572): Fix locking in cert_store_stats and get_ca_certs ([GH-114573](https://github.com/python/cpython/pull/114573))

* [gh-114572](https://github.com/python/cpython/issues/114572): Fix locking in cert_store_stats and get_ca_certs

cert_store_stats and get_ca_certs query the SSLContext's X509_STORE with
X509_STORE_get0_objects, but reading the result requires a lock. See
[openssl/openssl#23224](https://github.com/openssl/openssl/pull/23224) for details.

Instead, use X509_STORE_get1_objects, newly added in that PR.
X509_STORE_get1_objects does not exist in current OpenSSLs, but we can
polyfill it with X509_STORE_lock and X509_STORE_unlock.

* Work around const-correctness problem

* Add missing X509_STORE_get1_objects failure check

* Add blurb
(cherry picked from commit [bce6931](https://github.com/python/cpython/commit/bce693111bff906ccf9281c22371331aaff766ab))

Co-authored-by: David Benjamin <davidben@google.com>
```

[encukou](/encukou)
pushed a commit
that referenced
this issue
[Feb 29, 2024](#ref-commit-542f327)
[![@miss-islington](https://avatars.githubusercontent.com/u/31488909?s=40&u=31ad766412aaeb40a51582d882e34d67f2fae635&v=4)](/miss-islington) [![@davidben](https://avatars.githubusercontent.com/u/109690?s=40&v=4)](/davidben)

`[[3.12]](/python/cpython/commit/542f3272f56f31ed04e74c40635a913fbc12d286 "[3.12] gh-114572: Fix locking in cert_store_stats and get_ca_certs (GH-114573) (GH-115547)

gh-114572: Fix locking in cert_store_stats and get_ca_certs (GH-114573)

* gh-114572: Fix locking in cert_store_stats and get_ca_certs

cert_store_stats and get_ca_certs query the SSLContext's X509_STORE with
X509_STORE_get0_objects, but reading the result requires a lock. See
https://github.com/openssl/openssl/pull/23224 for details.

Instead, use X509_STORE_get1_objects, newly added in that PR.
X509_STORE_get1_objects does not exist in current OpenSSLs, but we can
polyfill it with X509_STORE_lock and X509_STORE_unlock.

* Work around const-correctness problem

* Add missing X509_STORE_get1_objects failure check

* Add blurb
(cherry picked from commit bce693111bff906ccf9281c22371331aaff766ab)

Co-authored-by: David Benjamin <davidben@google.com>") [gh-114572](https://github.com/python/cpython/issues/114572)[: Fix locking in cert_store_stats and get_ca_certs (](/python/cpython/commit/542f3272f56f31ed04e74c40635a913fbc12d286 "[3.12] gh-114572: Fix locking in cert_store_stats and get_ca_certs (GH-114573) (GH-115547)

gh-114572: Fix locking in cert_store_stats and get_ca_certs (GH-114573)

* gh-114572: Fix locking in cert_store_stats and get_ca_certs

cert_store_stats and get_ca_certs query the SSLContext's X509_STORE with
X509_STORE_get0_objects, but reading the result requires a lock. See
https://github.com/openssl/openssl/pull/23224 for details.

Instead, use X509_STORE_get1_objects, newly added in that PR.
X509_STORE_get1_objects does not exist in current OpenSSLs, but we can
polyfill it with X509_STORE_lock and X509_STORE_unlock.

* Work around const-correctness problem

* Add missing X509_STORE_get1_objects failure check

* Add blurb
(cherry picked from commit bce693111bff906ccf9281c22371331aaff766ab)

Co-authored-by: David Benjamin <davidben@google.com>")[G…](https://github.com/python/cpython/pull/114573)`
…

`[542f327](/python/cpython/commit/542f3272f56f31ed04e74c40635a913fbc12d286)`

```
[…H-114573](https://github.com/python/cpython/pull/114573)) ([GH-115547](https://github.com/python/cpython/pull/115547))

[gh-114572](https://github.com/python/cpython/issues/114572): Fix locking in cert_store_stats and get_ca_certs ([GH-114573](https://github.com/python/cpython/pull/114573))

* [gh-114572](https://github.com/python/cpython/issues/114572): Fix locking in cert_store_stats and get_ca_certs

cert_store_stats and get_ca_certs query the SSLContext's X509_STORE with
X509_STORE_get0_objects, but reading the result requires a lock. See
[openssl/openssl#23224](https://github.com/openssl/openssl/pull/23224) for details.

Instead, use X509_STORE_get1_objects, newly added in that PR.
X509_STORE_get1_objects does not exist in current OpenSSLs, but we can
polyfill it with X509_STORE_lock and X509_STORE_unlock.

* Work around const-correctness problem

* Add missing X509_STORE_get1_objects failure check

* Add blurb
(cherry picked from commit [bce6931](https://github.com/python/cpython/commit/bce693111bff906ccf9281c22371331aaff766ab))

Co-authored-by: David Benjamin <davidben@google.com>
```

[![@hugovk](https://avatars.githubusercontent.com/u/1324225?s=80&u=d7e2522cc357c1b8fed0f1c623c68c7331c70c56&v=4)](/hugovk)

Copy link

Member

### **[hugovk](/hugovk)** commented [Mar 15, 2024](#issuecomment-2000073857)

| Closing as the PR has been merged and backported. Thanks! |
| --- |

All reactions

Sorry, something went wrong.

[![@hugovk](https://avatars.githubusercontent.com/u/1324225?s=40&u=d7e2522cc357c1b8fed0f1c623c68c7331c70c56&v=4)](/hugovk)
[hugovk](/hugovk)
closed this as [completed](/python/cpython/issues?q=is%3Aissue+is%3Aclosed+archived%3Afalse+reason%3Acompleted)
[Mar 15, 2024](#event-12134429911)

[![@dongsupark](https://avatars.githubusercontent.com/u/10096906?s=40&v=4)](/dongsupark)
[dongsupark](/dongsupark)
mentioned this issue
[Mar 25, 2024](#ref-issue-2205959516)

[update: python
flatcar/Flatcar#1402](/flatcar/Flatcar/issues/1402)

Closed

[diegorusso](/diegorusso)
pushed a commit
to diegorusso/cpython
that referenced
this issue
[Apr 17, 2024](#ref-commit-eedb7c2)
[![@davidben](https://avatars.githubusercontent.com/u/109690?s=40&v=4)](/davidben) [![@diegorusso](https://avatars.githubusercontent.com/u/342108?s=40&v=4)](/diegorusso)

`[pythongh-114572](https://github.com/python/cpython/issues/114572)[: Fix locking in cert_store_stats and get_ca_certs (](/diegorusso/cpython/commit/eedb7c287517ea4163e3951fff13cfde100dc512 "gh-114572: Fix locking in cert_store_stats and get_ca_certs (#114573)

* gh-114572: Fix locking in cert_store_stats and get_ca_certs

cert_store_stats and get_ca_certs query the SSLContext's X509_STORE with
X509_STORE_get0_objects, but reading the result requires a lock. See
https://github.com/openssl/openssl/pull/23224 for details.

Instead, use X509_STORE_get1_objects, newly added in that PR.
X509_STORE_get1_objects does not exist in current OpenSSLs, but we can
polyfill it with X509_STORE_lock and X509_STORE_unlock.

* Work around const-correctness problem

* Add missing X509_STORE_get1_objects failure check

* Add blurb")[py…](https://github.com/python/cpython/pull/114573)`
…

`[eedb7c2](/diegorusso/cpython/commit/eedb7c287517ea4163e3951fff13cfde100dc512)`

```
[…thon#114573](https://github.com/python/cpython/pull/114573))

* [pythongh-114572](https://github.com/python/cpython/issues/114572): Fix locking in cert_store_stats and get_ca_certs

cert_store_stats and get_ca_certs query the SSLContext's X509_STORE with
X509_STORE_get0_objects, but reading the result requires a lock. See
[openssl/openssl#23224](https://github.com/openssl/openssl/pull/23224) for details.

Instead, use X509_STORE_get1_objects, newly added in that PR.
X509_STORE_get1_objects does not exist in current OpenSSLs, but we can
polyfill it with X509_STORE_lock and X509_STORE_unlock.

* Work around const-correctness problem

* Add missing X509_STORE_get1_objects failure check

* Add blurb
```

[sethmlarson](/sethmlarson)
added a commit
to sethmlarson/cpython
that referenced
this issue
[Apr 19, 2024](#ref-commit-732c7d5)
[![@sethmlarson](https://avatars.githubusercontent.com/u/18519037?s=40&u=41090cc65ae0a34aee49c7a35cfbd40e2e12eb53&v=4)](/sethmlarson)

`[[3.9]](/sethmlarson/cpython/commit/732c7d512e7cdf656a3f02a38c329b14a14a8573 "[3.9] gh-114572: Fix locking in cert_store_stats and get_ca_certs") [pythongh-114572](https://github.com/python/cpython/issues/114572)[: Fix locking in cert_store_stats and get_ca_certs](/sethmlarson/cpython/commit/732c7d512e7cdf656a3f02a38c329b14a14a8573 "[3.9] gh-114572: Fix locking in cert_store_stats and get_ca_certs")`

`[732c7d5](/sethmlarson/cpython/commit/732c7d512e7cdf656a3f02a38c329b14a14a8573)`

[![@bedevere-app](https://avatars.githubusercontent.com/in/388350?s=40&v=4)](/apps/bedevere-app)
[bedevere-app](/apps/bedevere-app)
bot
mentioned this issue
[Apr 19, 2024](#ref-pullrequest-2253630620)

[[3.9] gh-114572: Fix locking in cert\_store\_stats and get\_ca\_certs
#118109](/python/cpython/pull/118109)
 Merged

[sethmlarson](/sethmlarson)
added a commit
to sethmlarson/cpython
that referenced
this issue
[Apr 30, 2024](#ref-commit-e8268b1)
[![@sethmlarson](https://avatars.githubusercontent.com/u/18519037?s=40&u=41090cc65ae0a34aee49c7a35cfbd40e2e12eb53&v=4)](/sethmlarson)

`[[3.8] [3.9]](/sethmlarson/cpython/commit/e8268b14b12d8bde6bdf6b8cc2f676901a0d172a "[3.8] [3.9] gh-114572: Fix locking in cert_store_stats and get_ca_certs
(cherry picked from commit 732c7d512e7cdf656a3f02a38c329b14a14a8573)

Co-authored-by: Seth Michael Larson <seth@python.org>") [pythongh-114572](https://github.com/python/cpython/issues/114572)[: Fix locking in cert_store_stats and get_…](/sethmlarson/cpython/commit/e8268b14b12d8bde6bdf6b8cc2f676901a0d172a "[3.8] [3.9] gh-114572: Fix locking in cert_store_stats and get_ca_certs
(cherry picked from commit 732c7d512e7cdf656a3f02a38c329b14a14a8573)

Co-authored-by: Seth Michael Larson <seth@python.org>")`
…

`[e8268b1](/sethmlarson/cpython/commit/e8268b14b12d8bde6bdf6b8cc2f676901a0d172a)`

```
…ca_certs

(cherry picked from commit [732c7d5](https://github.com/sethmlarson/cpython/commit/732c7d512e7cdf656a3f02a38c329b14a14a8573))

Co-authored-by: Seth Michael Larson <seth@python.org>
```

[![@bedevere-app](https://avatars.githubusercontent.com/in/388350?s=40&v=4)](/apps/bedevere-app)
[bedevere-app](/apps/bedevere-app)
bot
mentioned this issue
[Apr 30, 2024](#ref-pullrequest-2272099966)

[[3.8] gh-114572: Fix locking in cert\_store\_stats and get\_ca\_certs
#118442](/python/cpython/pull/118442)
 Merged

[ambv](/ambv)
pushed a commit
that referenced
this issue
[May 10, 2024](#ref-commit-29c9728)
[![@sethmlarson](https://avatars.githubusercontent.com/u/18519037?s=40&u=41090cc65ae0a34aee49c7a35cfbd40e2e12eb53&v=4)](/sethmlarson)

`[[3.8]](/python/cpython/commit/29c97287d205bf2f410f4895ebce3f43b5160524 "[3.8] gh-114572: Fix locking in cert_store_stats and get_ca_certs (#118442)

(cherry picked from commit 732c7d512e7cdf656a3f02a38c329b14a14a8573)") [gh-114572](https://github.com/python/cpython/issues/114572)[: Fix locking in cert_store_stats and get_ca_certs (](/python/cpython/commit/29c97287d205bf2f410f4895ebce3f43b5160524 "[3.8] gh-114572: Fix locking in cert_store_stats and get_ca_certs (#118442)

(cherry picked from commit 732c7d512e7cdf656a3f02a38c329b14a14a8573)")[#1…](https://github.com/python/cpython/pull/118442)`
…

`[29c9728](/python/cpython/commit/29c97287d205bf2f410f4895ebce3f43b5160524)`

```
[…18442](https://github.com/python/cpython/pull/118442))

(cherry picked from commit [732c7d5](https://github.com/python/cpython/commit/732c7d512e7cdf656a3f02a38c329b14a14a8573))
```

[ambv](/ambv)
pushed a commit
that referenced
this issue
[May 10, 2024](#ref-commit-b228655)
[![@sethmlarson](https://avatars.githubusercontent.com/u/18519037?s=40&u=41090cc65ae0a34aee49c7a35cfbd40e2e12eb53&v=4)](/sethmlarson)

`[[3.9]](/python/cpython/commit/b228655c227b2ca298a8ffac44d14ce3d22f6faa "[3.9] gh-114572: Fix locking in cert_store_stats and get_ca_certs (#118109)") [gh-114572](https://github.com/python/cpython/issues/114572)[: Fix locking in cert_store_stats and get_ca_certs (](/python/cpython/commit/b228655c227b2ca298a8ffac44d14ce3d22f6faa "[3.9] gh-114572: Fix locking in cert_store_stats and get_ca_certs (#118109)")[#1…](https://github.com/python/cpython/pull/118109)`
…

`[b228655](/python/cpython/commit/b228655c227b2ca298a8ffac44d14ce3d22f6faa)`

```
[…18109](https://github.com/python/cpython/pull/118109))
```

[gentoo-bot](/gentoo-bot)
pushed a commit
to gentoo/cpython
that referenced
this issue
[May 21, 2024](#ref-commit-8ed9d5e)
[![@miss-islington](https://avatars.githubusercontent.com/u/31488909?s=40&u=31ad766412aaeb40a51582d882e34d67f2fae635&v=4)](/miss-islington) [![@davidben](https://avatars.githubusercontent.com/u/109690?s=40&v=4)](/davidben) [![@mgorny](https://avatars.githubusercontent.com/u/110765?s=40&v=4)](/mgorny)

`[[3.10]](/gentoo/cpython/commit/8ed9d5eb1471774894eb47ee63646917bf5e3689 "[3.10] gh-114572: Fix locking in cert_store_stats and get_ca_certs (GH-114573) (#115548)

gh-114572: Fix locking in cert_store_stats and get_ca_certs (GH-114573)

* gh-114572: Fix locking in cert_store_stats and get_ca_certs

cert_store_stats and get_ca_certs query the SSLContext's X509_STORE with
X509_STORE_get0_objects, but reading the result requires a lock. See
https://github.com/openssl/openssl/pull/23224 for details.

Instead, use X509_STORE_get1_objects, newly added in that PR.
X509_STORE_get1_objects does not exist in current OpenSSLs, but we can
polyfill it with X509_STORE_lock and X509_STORE_unlock.

* Work around const-correctness problem

* Add missing X509_STORE_get1_objects failure check

* Add blurb
(cherry picked from commit bce693111bff906ccf9281c22371331aaff766ab)

Co-authored-by: David Benjamin <davidben@google.com>") [pythongh-114572](https://github.com/python/cpython/issues/114572)[: Fix locking in cert_store_stats and get_ca_ce…](/gentoo/cpython/commit/8ed9d5eb1471774894eb47ee63646917bf5e3689 "[3.10] gh-114572: Fix locking in cert_store_stats and get_ca_certs (GH-114573) (#115548)

gh-114572: Fix locking in cert_store_stats and get_ca_certs (GH-114573)

* gh-114572: Fix locking in cert_store_stats and get_ca_certs

cert_store_stats and get_ca_certs query the SSLContext's X509_STORE with
X509_STORE_get0_objects, but reading the result requires a lock. See
https://github.com/openssl/openssl/pull/23224 for details.

Instead, use X509_STORE_get1_objects, newly added in that PR.
X509_STORE_get1_objects does not exist in current OpenSSLs, but we can
polyfill it with X509_STORE_lock and X509_STORE_unlock.

* Work around const-correctness problem

* Add missing X509_STORE_get1_objects failure check

* Add blurb
(cherry picked from commit bce693111bff906ccf9281c22371331aaff766ab)

Co-authored-by: David Benjamin <davidben@google.com>")`
…

`[8ed9d5e](/gentoo/cpython/commit/8ed9d5eb1471774894eb47ee63646917bf5e3689)`

```
…rts ([pythonGH-114573](https://github.com/python/cpython/pull/114573)) ([python#115548](https://github.com/python/cpython/pull/115548))

[pythongh-114572](https://github.com/python/cpython/issues/114572): Fix locking in cert_store_stats and get_ca_certs ([pythonGH-114573](https://github.com/python/cpython/pull/114573))

* [pythongh-114572](https://github.com/python/cpython/issues/114572): Fix locking in cert_store_stats and get_ca_certs

cert_store_stats and get_ca_certs query the SSLContext's X509_STORE with
X509_STORE_get0_objects, but reading the result requires a lock. See
[openssl/openssl#23224](https://github.com/openssl/openssl/pull/23224) for details.

Instead, use X509_STORE_get1_objects, newly added in that PR.
X509_STORE_get1_objects does not exist in current OpenSSLs, but we can
polyfill it with X509_STORE_lock and X509_STORE_unlock.

* Work around const-correctness problem

* Add missing X509_STORE_get1_objects failure check

* Add blurb
(cherry picked from commit [bce6931](https://github.com/gentoo/cpython/commit/bce693111bff906ccf9281c22371331aaff766ab))

Co-authored-by: David Benjamin <davidben@google.com>
```

[gentoo-bot](/gentoo-bot)
pushed a commit
to gentoo/cpython
that referenced
this issue
[May 21, 2024](#ref-commit-a6a90ca)
[![@miss-islington](https://avatars.githubusercontent.com/u/31488909?s=40&u=31ad766412aaeb40a51582d882e34d67f2fae635&v=4)](/miss-islington) [![@davidben](https://avatars.githubusercontent.com/u/109690?s=40&v=4)](/davidben) [![@mgorny](https://avatars.githubusercontent.com/u/110765?s=40&v=4)](/mgorny)

`[[3.10]](/gentoo/cpython/commit/a6a90cac7e1af91b032dcf0df13437857bc6c112 "[3.10] gh-114572: Fix locking in cert_store_stats and get_ca_certs (GH-114573) (#115548)

gh-114572: Fix locking in cert_store_stats and get_ca_certs (GH-114573)

* gh-114572: Fix locking in cert_store_stats and get_ca_certs

cert_store_stats and get_ca_certs query the SSLContext's X509_STORE with
X509_STORE_get0_objects, but reading the result requires a lock. See
https://github.com/openssl/openssl/pull/23224 for details.

Instead, use X509_STORE_get1_objects, newly added in that PR.
X509_STORE_get1_objects does not exist in current OpenSSLs, but we can
polyfill it with X509_STORE_lock and X509_STORE_unlock.

* Work around const-correctness problem

* Add missing X509_STORE_get1_objects failure check

* Add blurb
(cherry picked from commit bce693111bff906ccf9281c22371331aaff766ab)

Co-authored-by: David Benjamin <davidben@google.com>") [pythongh-114572](https://github.com/python/cpython/issues/114572)[: Fix locking in cert_store_stats and get_ca_ce…](/gentoo/cpython/commit/a6a90cac7e1af91b032dcf0df13437857bc6c112 "[3.10] gh-114572: Fix locking in cert_store_stats and get_ca_certs (GH-114573) (#115548)

gh-114572: Fix locking in cert_store_stats and get_ca_certs (GH-114573)

* gh-114572: Fix locking in cert_store_stats and get_ca_certs

cert_store_stats and get_ca_certs query the SSLContext's X509_STORE with
X509_STORE_get0_objects, but reading the result requires a lock. See
https://github.com/openssl/openssl/pull/23224 for details.

Instead, use X509_STORE_get1_objects, newly added in that PR.
X509_STORE_get1_objects does not exist in current OpenSSLs, but we can
polyfill it with X509_STORE_lock and X509_STORE_unlock.

* Work around const-correctness problem

* Add missing X509_STORE_get1_objects failure check

* Add blurb
(cherry picked from commit bce693111bff906ccf9281c22371331aaff766ab)

Co-authored-by: David Benjamin <davidben@google.com>")`
…

`[a6a90ca](/gentoo/cpython/commit/a6a90cac7e1af91b032dcf0df13437857bc6c112)`

```
…rts ([pythonGH-114573](https://github.com/python/cpython/pull/114573)) ([python#115548](https://github.com/python/cpython/pull/115548))

[pythongh-114572](https://github.com/python/cpython/issues/114572): Fix locking in cert_store_stats and get_ca_certs ([pythonGH-114573](https://github.com/python/cpython/pull/114573))

* [pythongh-114572](https://github.com/python/cpython/issues/114572): Fix locking in cert_store_stats and get_ca_certs

cert_store_stats and get_ca_certs query the SSLContext's X509_STORE with
X509_STORE_get0_objects, but reading the result requires a lock. See
[openssl/openssl#23224](https://github.com/openssl/openssl/pull/23224) for details.

Instead, use X509_STORE_get1_objects, newly added in that PR.
X509_STORE_get1_objects does not exist in current OpenSSLs, but we can
polyfill it with X509_STORE_lock and X509_STORE_unlock.

* Work around const-correctness problem

* Add missing X509_STORE_get1_objects failure check

* Add blurb
(cherry picked from commit [bce6931](https://github.com/gentoo/cpython/commit/bce693111bff906ccf9281c22371331aaff766ab))

Co-authored-by: David Benjamin <davidben@google.com>
```

[mcepl](/mcepl)
pushed a commit
to openSUSE-Python/cpython
that referenced
this issue
[Jun 21, 2024](#ref-commit-e701181)
[![@sethmlarson](https://avatars.githubusercontent.com/u/18519037?s=40&u=41090cc65ae0a34aee49c7a35cfbd40e2e12eb53&v=4)](/sethmlarson) [![@mcepl](https://avatars.githubusercontent.com/u/198999?s=40&v=4)](/mcepl)

`[[](/openSUSE-Python/cpython/commit/e7011813a2b8f078182d29c252ea728b1ef163d8 "[CVE-2024-0397] Fix locking in cert_store_stats and get_ca_certs

cert_store_stats and get_ca_certs query the SSLContext's
X509_STORE with X509_STORE_get0_objects, but reading the result
requires a lock. See gh#openssl/openssl#23224 for details.

Instead, use X509_STORE_get1_objects, newly added in that
PR. X509_STORE_get1_objects does not exist in current
OpenSSLs, but we can polyfill it with X509_STORE_lock and
X509_STORE_unlock.

From-PR: gh#python/cpython!114573
Fixes: gh#python/cpython#114572
Fixes: bsc#1226447 (CVE-2024-0397)
Patch: CVE-2024-0397-memrace_ssl.SSLContext_cert_store.patch")[CVE-2024-0397](https://github.com/advisories/GHSA-xhf3-pp4q-gxh5 "CVE-2024-0397")[] Fix locking in cert_store_stats and get_ca_certs](/openSUSE-Python/cpython/commit/e7011813a2b8f078182d29c252ea728b1ef163d8 "[CVE-2024-0397] Fix locking in cert_store_stats and get_ca_certs

cert_store_stats and get_ca_certs query the SSLContext's
X509_STORE with X509_STORE_get0_objects, but reading the result
requires a lock. See gh#openssl/openssl#23224 for details.

Instead, use X509_STORE_get1_objects, newly added in that
PR. X509_STORE_get1_objects does not exist in current
OpenSSLs, but we can polyfill it with X509_STORE_lock and
X509_STORE_unlock.

From-PR: gh#python/cpython!114573
Fixes: gh#python/cpython#114572
Fixes: bsc#1226447 (CVE-2024-0397)
Patch: CVE-2024-0397-memrace_ssl.SSLContext_cert_store.patch")`
…

`[e701181](/openSUSE-Python/cpython/commit/e7011813a2b8f078182d29c252ea728b1ef163d8)`

```
cert_store_stats and get_ca_certs query the SSLContext's
X509_STORE with X509_STORE_get0_objects, but reading the result
requires a lock. See gh#[openssl/openssl#23224](https://github.com/openssl/openssl/pull/23224) for details.

Instead, use X509_STORE_get1_objects, newly added in that
PR. X509_STORE_get1_objects does not exist in current
OpenSSLs, but we can polyfill it with X509_STORE_lock and
X509_STORE_unlock.

From-PR: gh#python/cpython!114573
Fixes: gh#[python#114572](https://github.com/python/cpython/issues/114572)
Fixes: bsc#1226447 ([CVE-2024-0397](https://github.com/advisories/GHSA-xhf3-pp4q-gxh5 "CVE-2024-0397"))
Patch: [CVE-2024-0397](https://github.com/advisories/GHSA-xhf3-pp4q-gxh5 "CVE-2024-0397")-memrace_ssl.SSLContext_cert_store.patch
```

[rickprice](/rickprice)
added a commit
to ActiveState/cpython
that referenced
this issue
[Jun 27, 2024](#ref-commit-eccf2f4)
[![@rickprice](https://avatars.githubusercontent.com/u/1013454?s=40&u=0ee5f0661235b463155f59dd127e785840c2813d&v=4)](/rickprice)

`[CVE-2024-0397](https://github.com/advisories/GHSA-xhf3-pp4q-gxh5 "CVE-2024-0397") [Fix for the problem, backported from Python3.8](/ActiveState/cpython/commit/eccf2f4989f79ca0b42f30ec75b634ac6f449cec "CVE-2024-0397 Fix for the problem, backported from Python3.8 gh-114572") [pythongh…](https://github.com/python/cpython/issues/114572)`
…

`[eccf2f4](/ActiveState/cpython/commit/eccf2f4989f79ca0b42f30ec75b634ac6f449cec)`

```
[…-114572](https://github.com/python/cpython/issues/114572)
```

[rickprice](/rickprice)
added a commit
to ActiveState/cpython
that referenced
this issue
[Jun 27, 2024](#ref-commit-561a5de)
[![@rickprice](https://avatars.githubusercontent.com/u/1013454?s=40&u=0ee5f0661235b463155f59dd127e785840c2813d&v=4)](/rickprice)

`[CVE-2024-0397](https://github.com/advisories/GHSA-xhf3-pp4q-gxh5 "CVE-2024-0397") [Fix locking in cert_store_stats and get_ca_certs](/ActiveState/cpython/commit/561a5de331a93515f61c6e770d245e6a57212121 "CVE-2024-0397 Fix locking in cert_store_stats and get_ca_certs

Backported from Python3.8 gh-114572")`
…

`[561a5de](/ActiveState/cpython/commit/561a5de331a93515f61c6e770d245e6a57212121)`

```
Backported from Python3.8 [pythongh-114572](https://github.com/python/cpython/issues/114572)
```

[rickprice](/rickprice)
added a commit
to ActiveState/cpython
that referenced
this issue
[Jun 27, 2024](#ref-commit-7a4117c)
[![@rickprice](https://avatars.githubusercontent.com/u/1013454?s=40&u=0ee5f0661235b463155f59dd127e785840c2813d&v=4)](/rickprice)

`[CVE-2024-0397](https://github.com/advisories/GHSA-xhf3-pp4q-gxh5 "CVE-2024-0397") [Fix locking in cert_store_stats and get_ca_certs](/ActiveState/cpython/commit/7a4117c68c61756dd5582ac97c423d644b53324e "CVE-2024-0397 Fix locking in cert_store_stats and get_ca_certs

Backported from : [3.8] gh-114572: Fix locking in cert_store_stats and get_ca_certs #118442")`
…

`[7a4117c](/ActiveState/cpython/commit/7a4117c68c61756dd5582ac97c423d644b53324e)`

```
Backported from : [3.8] [pythongh-114572](https://github.com/python/cpython/issues/114572): Fix locking in cert_store_stats and get_ca_certs [python#118442](https://github.com/python/cpython/pull/118442)
```

[![@icanhasmath](https://avatars.githubusercontent.com/u/43051639?s=40&v=4)](/icanhasmath)
[icanhasmath](/icanhasmath)
mentioned this issue
[Jun 27, 2024](#ref-pullrequest-2378680506)

[CVE-2024-0397 Fix for the problem, backported from Python3.8 gh-114572
ActiveState/cpython#53](/ActiveState/cpython/pull/53)
 Merged

[icanhasmath](/icanhasmath)
added a commit
to ActiveState/cpython
that referenced
this issue
[Jun 27, 2024](#ref-commit-43d80c0)
[![@icanhasmath](https://avatars.githubusercontent.com/u/43051639?s=40&v=4)](/icanhasmath)

`[Merge pull request](/ActiveState/cpython/commit/43d80c0523f6f61c1c5e0050574d55fe655b85bf "Merge pull request #53 from ActiveState/BE-4053-cve-2024-0397

CVE-2024-0397 Fix for the problem, backported from Python3.8 gh-114572") [#53](https://github.com/ActiveState/cpython/pull/53) [from ActiveState/BE-4053-](/ActiveState/cpython/commit/43d80c0523f6f61c1c5e0050574d55fe655b85bf "Merge pull request #53 from ActiveState/BE-4053-cve-2024-0397

CVE-2024-0397 Fix for the problem, backported from Python3.8 gh-114572")[cve-2024-0397](https://github.com/advisories/GHSA-xhf3-pp4q-gxh5 "CVE-2024-0397")`
…

`[43d80c0](/ActiveState/cpython/commit/43d80c0523f6f61c1c5e0050574d55fe655b85bf)`

```
[CVE-2024-0397](https://github.com/advisories/GHSA-xhf3-pp4q-gxh5 "CVE-2024-0397") Fix for the problem, backported from Python3.8 [pythongh-114572](https://github.com/python/cpython/issues/114572)
```

[icanhasmath](/icanhasmath)
added a commit
to ActiveState/cpython
that referenced
this issue
[Jun 27, 2024](#ref-commit-5cb556a)
[![@icanhasmath](https://avatars.githubusercontent.com/u/43051639?s=40&v=4)](/icanhasmath)

`[Revert "](/ActiveState/cpython/commit/5cb556a5766149f89f67cbeaf331f4328a24fe6f "Revert \"CVE-2024-0397 Fix for the problem, backported from Python3.8 gh-114572\"")[CVE-2024-0397](https://github.com/advisories/GHSA-xhf3-pp4q-gxh5 "CVE-2024-0397") [Fix for the problem, backported from Python3.8](/ActiveState/cpython/commit/5cb556a5766149f89f67cbeaf331f4328a24fe6f "Revert \"CVE-2024-0397 Fix for the problem, backported from Python3.8 gh-114572\"") [p…](https://github.com/python/cpython/issues/114572)`
…

`[5cb556a](/ActiveState/cpython/commit/5cb556a5766149f89f67cbeaf331f4328a24fe6f)`

```
[…ythongh-114572](https://github.com/python/cpython/issues/114572)"
```

[rickprice](/rickprice)
added a commit
to ActiveState/cpython
that referenced
this issue
[Jun 28, 2024](#ref-commit-0dd4acb)
[![@rickprice](https://avatars.githubusercontent.com/u/1013454?s=40&u=0ee5f0661235b463155f59dd127e785840c2813d&v=4)](/rickprice)

`[CVE-2024-0397](https://github.com/advisories/GHSA-xhf3-pp4q-gxh5 "CVE-2024-0397")`
…

`[0dd4acb](/ActiveState/cpython/commit/0dd4acb5689c79fb6f4fbe7825d26b202010c7ae)`

```
Backport
[python@e8268b1](https://github.com/python/cpython/commit/e8268b14b12d8bde6bdf6b8cc2f676901a0d172a)
Which is [pythongh-114572](https://github.com/python/cpython/issues/114572)
```

[rickprice](/rickprice)
added a commit
to ActiveState/cpython
that referenced
this issue
[Jun 28, 2024](#ref-commit-04d4542)
[![@rickprice](https://avatars.githubusercontent.com/u/1013454?s=40&u=0ee5f0661235b463155f59dd127e785840c2813d&v=4)](/rickprice)

`[CVE-2024-0397](https://github.com/advisories/GHSA-xhf3-pp4q-gxh5 "CVE-2024-0397")`
…

`[04d4542](/ActiveState/cpython/commit/04d4542969fb9aaa2aa3738736d2b5de10507850)`

```
Backport
[python@e8268b1](https://github.com/python/cpython/commit/e8268b14b12d8bde6bdf6b8cc2f676901a0d172a)
Which is [pythongh-114572](https://github.com/python/cpython/issues/114572)
```

[rickprice](/rickprice)
added a commit
to ActiveState/cpython
that referenced
this issue
[Jun 29, 2024](#ref-commit-56d38c9)
[![@rickprice](https://avatars.githubusercontent.com/u/1013454?s=40&u=0ee5f0661235b463155f59dd127e785840c2813d&v=4)](/rickprice)

`[CVE-2024-0397](https://github.com/advisories/GHSA-xhf3-pp4q-gxh5 "CVE-2024-0397")`
…

`[56d38c9](/ActiveState/cpython/commit/56d38c9af80d12e1186f6491413f7c93c0a27f80)`

```
Backport
[python@e8268b1](https://github.com/python/cpython/commit/e8268b14b12d8bde6bdf6b8cc2f676901a0d172a)
Which is [pythongh-114572](https://github.com/python/cpython/issues/114572)
```

[rickprice](/rickprice)
added a commit
to ActiveState/cpython
that referenced
this issue
[Jun 29, 2024](#ref-commit-01a9003)
[![@rickprice](https://avatars.githubusercontent.com/u/1013454?s=40&u=0ee5f0661235b463155f59dd127e785840c2813d&v=4)](/rickprice)

`[CVE-2024-0397](https://github.com/advisories/GHSA-xhf3-pp4q-gxh5 "CVE-2024-0397") [Fix locking in cert_store_stats and get_ca_certs](/ActiveState/cpython/commit/01a9003ad04a0c9cb4f4d5bc1a7ba513d31b29e3 "CVE-2024-0397 Fix locking in cert_store_stats and get_ca_certs

Backported from : [3.8] gh-114572: Fix locking in cert_store_stats and
get_ca_certs #118442")`
…

`[01a9003](/ActiveState/cpython/commit/01a9003ad04a0c9cb4f4d5bc1a7ba513d31b29e3)`

```
Backported from : [3.8] [pythongh-114572](https://github.com/python/cpython/issues/114572): Fix locking in cert_store_stats and
get_ca_certs [python#118442](https://github.com/python/cpython/pull/118442)
```

[![@rickprice](https://avatars.githubusercontent.com/u/1013454?s=40&u=0ee5f0661235b463155f59dd127e785840c2813d&v=4)](/rickprice)
[rickprice](/rickprice)
mentioned this issue
[Jun 29, 2024](#ref-pullrequest-2382011050)

[CVE-2024-0397 Fix locking in cert\_store\_stats and get\_ca\_certs
ActiveState/cpython#56](/ActiveState/cpython/pull/56)
 Merged

[rickprice](/rickprice)
added a commit
to ActiveState/cpython
that referenced
this issue
[Jul 12, 2024](#ref-commit-9aa2030)
[![@rickprice](https://avatars.githubusercontent.com/u/1013454?s=40&u=0ee5f0661235b463155f59dd127e785840c2813d&v=4)](/rickprice)

`[CVE-2024-0397](https://github.com/advisories/GHSA-xhf3-pp4q-gxh5 "CVE-2024-0397") [Fix locking in cert_store_stats and get_ca_certs](/ActiveState/cpython/commit/9aa203051d4e3698c3102151ba629e04503de8f5 "CVE-2024-0397 Fix locking in cert_store_stats and get_ca_certs

Backported from : [3.8] gh-114572: Fix locking in cert_store_stats and
get_ca_certs #118442")`
…

`[9aa2030](/ActiveState/cpython/commit/9aa203051d4e3698c3102151ba629e04503de8f5)`

```
Backported from : [3.8] [pythongh-114572](https://github.com/python/cpython/issues/114572): Fix locking in cert_store_stats and
get_ca_certs [python#118442](https://github.com/python/cpython/pull/118442)
```

[icanhasmath](/icanhasmath)
pushed a commit
to ActiveState/cpython
that referenced
this issue
[Aug 8, 2024](#ref-commit-ea1d3ab)
[![@rickprice](https://avatars.githubusercontent.com/u/1013454?s=40&u=0ee5f0661235b463155f59dd127e785840c2813d&v=4)](/rickprice) [![@icanhasmath](https://avatars.githubusercontent.com/u/43051639?s=40&v=4)](/icanhasmath)

`[CVE-2024-0397](https://github.com/advisories/GHSA-xhf3-pp4q-gxh5 "CVE-2024-0397") [Fix locking in cert_store_stats and get_ca_certs](/ActiveState/cpython/commit/ea1d3abea1e985e191db45b2cf80d438765ed0d2 "CVE-2024-0397 Fix locking in cert_store_stats and get_ca_certs

Backported from : [3.8] gh-114572: Fix locking in cert_store_stats and get_ca_certs #118442")`
…

`[ea1d3ab](/ActiveState/cpython/commit/ea1d3abea1e985e191db45b2cf80d438765ed0d2)`

```
Backported from : [3.8] [pythongh-114572](https://github.com/python/cpython/issues/114572): Fix locking in cert_store_stats and get_ca_certs [python#118442](https://github.com/python/cpython/pull/118442)
```

[![@fcollman](https://avatars.githubusercontent.com/u/782341?s=40&v=4)](/fcollman)
[fcollman](/fcollman)
mentioned this issue
[Oct 22, 2024](#ref-issue-2539085159)

[urllib3 version 2 incompatibility
seung-lab/cloud-volume#636](/seung-lab/cloud-volume/issues/636)

Open

[Sign up for free](/join?source=comment-repo)
**to join this conversation on GitHub**.
Already have an account?
[Sign in to comment](/login?return_to=https%3A%2F%2Fgithub.com%2Fpython%2Fcpython%2Fissues%2F114572)

Assignees

No one assigned

Labels

[type-bug](/python/cpython/labels/type-bug)
An unexpected behavior, bug, or error

Projects

None yet

Milestone

No milestone

Development

No branches or pull requests

2 participants

[![@davidben](https://avatars.githubusercontent.com/u/109690?s=52&v=4)](/davidben) [![@hugovk](https://avatars.githubusercontent.com/u/1324225?s=52&v=4)](/hugovk)

## Footer

© 2025 GitHub, Inc.

### Footer navigation

* [Terms](https://docs.github.com/site-policy/github-terms/github-terms-of-service)
* [Privacy](https://docs.github.com/site-policy/privacy-policies/github-privacy-statement)
* [Security](https://github.com/security)
* [Status](https://www.githubstatus.com/)
* [Docs](https://docs.github.com/)
* [Contact](https://support.github.com?tags=dotcom-footer)
* Manage cookies
* Do not share my personal information

You can’t perform that action at this time.

